File: neetcode_150/minimumWindowSubstring.java
import java.util.HashMap;

public class minimumWindowSubstring {
    public static void main(String[] args) {

    }
    public String minWindow(String s, String t) {
        if(t.length()>s.length()){
            return "";
        }
        HashMap<Character, Integer> map = new HashMap();
        for(int i =0; i<t.length(); i++){
            map.put(t.charAt(i), map.getOrDefault(t.charAt(i),0)+1);
        }
        int left = 0, right = 0;
        int counter = 0;
        int length = Integer.MAX_VALUE;
        int start = -1;
        while(right<s.length()){
            if(map.containsKey(s.charAt(right)) && map.get(s.charAt(right))>0){
                counter++;
            }
            map.put(s.charAt(right), map.get(s.charAt(right))-1);
            while(counter == t.length()){
                if(right-left+1<length){
                    length = right-left+1;
                    start = left;
                }
                map.put(s.charAt(left), map.get(s.charAt(left))+1);
                if(map.get(s.charAt(left))>0){
                    counter--;
                }
                left++;
            }
            right++;
        }
        if (start == -1) {
            return "";
        }
        return s.substring(start, start+length);
    }
}

File: neetcode_150/minimumWindowSubstring.py
from collections import Counter

def minWindow(self, s: str, t: str) -> str:
    if len(t) > len(s):
        return ""
    if s == t:
        return s
    freq = Counter(t)
    window = {}
    left, start = 0, -1
    have, need = 0, len(freq)
    length = float('inf')
    for right in range(len(s)):
        curr = s[right]
        window[curr] = window.get(window[curr],0)+1
        if curr in freq and window[curr] == freq[curr]:
            have+=1
        while have == need:
            temp_len = right-left+1
            if temp_len < length:
                start = left
                length = temp_len
            leftChar = s[left]
            window[leftChar] -= 1
            if leftChar in freq and window[leftChar] < freq[leftChar]:
                have-=1
            left+=1
    if start == -1:
        return ""
    return s[start:start+length]

